### ISSUE TITLE
proof of correctness of your check in the busy loop

### ISSUE THREAD

**Issue by simonzhang00** [giotto-ai/giotto-ph#27]
There is a check that the pivot_column_index entry has not changed in a busy do...while loop. Can you prove the correctness of this?

**Comment by simonzhang00** [comment]
Yes, it suffices to check that `piv` does not change in between the two assignments `piv := pivots[l]` and `piv_column := R[piv]` so that the index of `piv_column` is actually `piv` later in execution. The check `low(piv_column)==l` is more expensive. Notice that `pivots[l]` is monotonically moving toward the left over execution so that if you check `pivots[l]` again, either it stays the same or it has moved left (it never comes back). Thus you cannot have inconsistent `piv_column` and `piv` and have the second read on `pivots[l]` equal `piv`. This explains why you have the condition in the while [loop](https://github.com/giotto-ai/giotto-ph/blob/1d0c628fe5e5c5c4f712bbc1fb9ddc3acc347f9d/gph/src/ripser.h#L1239) and this [line](https://github.com/giotto-ai/giotto-ph/blob/1d0c628fe5e5c5c4f712bbc1fb9ddc3acc347f9d/gph/src/ripser.h#L1237) checking `pivots[l]` again after the second assignment.

**Comment by ulupo** [comment]
@simonzhang00 wondering if you have any more comments on this topic?  I think it would be interesting to see if your point can lead to a speedup (e.g. as @MonkeyBreaker is proposing).

**Comment by simonzhang00** [comment]
I have no further comments. Thanks and good luck.
